src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,139 @@
 package com.es.phoneshop.model.product;
 
-import java.util.List;
+import java.util.*;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
-    @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+  private static volatile ArrayListProductDao instance;
+  private ReadWriteLock lock;
+  private List<Product> products;
+  private long maxId;
+
+  private ArrayListProductDao() {
+    this.maxId = 1;
+    this.products = new ArrayList<>();
+    this.lock = new ReentrantReadWriteLock();
+  }
+
+  public static ArrayListProductDao getInstance() {
+    ArrayListProductDao localInstance = instance;
+    if (localInstance == null) {
+      synchronized (ArrayListProductDao.class) {
+        localInstance = instance;
+        if (localInstance == null) {
+          localInstance = new ArrayListProductDao();
+          instance = localInstance;
+        }
+      }
+    }
+    return localInstance;
+  }
+
+  @Override
+  public Product getProduct(Long id) throws ProductNotFoundException {
+    Product result;
+    lock.readLock().lock();
+    try {
+      result = products.stream()
+              .filter(product -> product.getId().equals(id))
+              .findAny()
+              .orElseThrow(() -> new ProductNotFoundException(String.valueOf(id)));
+    } finally {
+      lock.readLock().unlock();
+    }
+    return result;
+  }
+
+  @Override
+  public List<Product> findProducts(String query, SortField sortField, SortOrder sortOrder) {
+    List<Product> productList;
+    lock.readLock().lock();
+    try {
+      String[] queryWords = (query == null || query.trim().isEmpty())
+              ? null
+              : query.trim().toLowerCase(Locale.ROOT).split("\\s");
+      Comparator<Product> productComparator = getSortOrderComparator(getSortFieldComparator(queryWords, sortField),
+              sortOrder);
+      productList = products.stream()
+              .filter(p -> p.getPrice() != null)
+              .filter(p -> p.getStock() > 0)
+              .filter(product -> queryWords == null
+                      || Arrays.stream(queryWords).anyMatch(
+                              s -> product.getDescription().toLowerCase(Locale.ROOT).contains(s)))
```

Comment:
Can we extract these conditions in a separate method?

```
queryWords == null
                      || Arrays.stream(queryWords).anyMatch(
                              s -> product.getDescription().toLowerCase(Locale.ROOT).contains(s))
```

so the code is something like the following, which is more readable:

`filter(product -> shouldDisplayProduct(queryWords, product))`